pkg/controlplane/control/peer.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package control
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/clusterlink-net/clusterlink/pkg/apis/clusterlink.net/v1alpha1"
+	"github.com/clusterlink-net/clusterlink/pkg/controlplane/peer"
+	"github.com/clusterlink-net/clusterlink/pkg/util/tls"
+)
+
+const (
+	heartbeatInterval = 10 * time.Second
+)
+
+// peerMonitor monitors a single peer.
+type peerMonitor struct {
+	lock           sync.Mutex
+	pr             *v1alpha1.Peer
+	client         *peer.Client
+	statusCallback func(*v1alpha1.Peer)
+
+	wg     *sync.WaitGroup
+	stopCh chan struct{}
+
+	logger *logrus.Entry
+}
+
+// peerManager manages peers status.
+type peerManager struct {
+	client         client.Client
+	peerTLS        *tls.ParsedCertData
+	statusCallback func(*v1alpha1.Peer)
+
+	lock     sync.Mutex
+	monitors map[string]*peerMonitor
+
+	stopped         bool
+	monitorWG       sync.WaitGroup
+	updaterWG       sync.WaitGroup
+	stopCh          chan struct{}
+	statusUpdatesCh chan *v1alpha1.Peer
+
+	logger *logrus.Entry
+}
+
+func (m *peerMonitor) Peer() v1alpha1.Peer {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return *m.pr
+}
+
+func (m *peerMonitor) Start() {
+	defer m.wg.Done()
+
+	ticker := time.NewTicker(heartbeatInterval)
+	defer ticker.Stop()
+
+	backoffConfig := backoff.NewExponentialBackOff()
+
+	reachable := false
+	reachableCond := metav1.Condition{
+		Type:   v1alpha1.PeerReachable,
+		Status: metav1.ConditionFalse,
+		Reason: "Heartbeat",
+	}
+
+	for {
+		select {
+		case <-m.stopCh:
+			return
+		default:
+			break
+		}
+
+		err := backoff.Retry(m.client.GetHeartbeat, backoffConfig)
+		if heartbeatOK := err == nil; heartbeatOK != reachable {
+			m.logger.Infof("Heartbeat result: %v", heartbeatOK)
+
+			if heartbeatOK {
+				reachableCond.Status = metav1.ConditionTrue
+				backoffConfig.MaxElapsedTime = heartbeatInterval
+			} else {
+				reachableCond.Status = metav1.ConditionFalse
+				backoffConfig.MaxElapsedTime = 0
+			}
+
+			reachable = heartbeatOK
+
+			m.lock.Lock()
+			meta.SetStatusCondition(&m.pr.Status.Conditions, reachableCond)
+			m.lock.Unlock()
+
+			m.statusCallback(m.pr)
```

Comment:
Can you add a comment why you call the callback

Reply:
Sure (it's for non-CRD mode, which does not watch Peer status).